thread_pthread_mn.c

```diff
@@ -674,12 +713,30 @@ timer_thread_register_waiting(rb_thread_t *th, int fd, enum thread_sched_waiting
 }
 
 static void
-timer_thread_unregister_waiting(rb_thread_t *th, int fd)
+timer_thread_unregister_waiting(rb_thread_t *th, int fd, enum thread_sched_waiting_flag flags)
 {
     RUBY_DEBUG_LOG("th:%u fd:%d", rb_th_serial(th), fd);
-
+#if HAVE_SYS_EVENT_H
+    if (flags) {
+        struct kevent ke[2];
+        int num_events = 0;
+
+        if (flags & thread_sched_waiting_io_read) {
+            EV_SET(&ke[num_events], fd, EVFILT_READ, EV_DELETE, 0, 0, NULL);
+            num_events++;
+        }
+        if (flags & thread_sched_waiting_io_write) {
+            EV_SET(&ke[num_events], fd, EVFILT_WRITE, EV_DELETE, 0, 0, NULL);
+            num_events++;
+        }
+        if (kevent(timer_th.kq_fd, ke, num_events, NULL, 0, NULL) == -1) {
+            perror("kevent");
+            rb_bug("unregister/kevent fails. errno:%d", errno);
+        }
+    }
+#else
     // Linux 2.6.9 or later is needed to pass NULL as data.
-    if (epoll_ctl(timer_th.epoll_fd, EPOLL_CTL_DEL, fd, NULL) == -1) {
+    if (epoll_ctl(timer_th.kq_fd, EPOLL_CTL_DEL, fd, NULL) == -1) {
```

Comment:
I was somewhat worried this may bring confusion 😕

I renamed the existing `epoll_fd` to `kq_fd`. Since both epoll and kqueue are **k**ernel event **q**ueues, it felt reasonable as a name.

But because it sounds so much like `kqueue`, maybe it's confusing. Since they're both `int` descriptors, it allowed me to avoid some `#if` checks by having a shared name. So I could:

1. Change the name to something less overloaded sounding? `event_fd`?
2. Keep both and just have one or two additional checks.
3. Leave it?

Reply:
I changed it to `event_fd`